GSOC-2021 Project Page for Git Credentials Binding (#4516)
jenkins-infra/jenkins.io: content/projects/gsoc/2021/projects/git-credentials-binding.adoc

Users will be able to run authenticated git commands in their Jenkins Pipelines without modifying the git plugin.

=== Implementation

> Both `gitUsernamePassword` and `gitSshPrivateKey` bindings depends on the https://plugins.jenkins.io/credentials/[Credential Plugin]

Reviewer: ```suggestion
Both `gitUsernamePassword` and `gitSshPrivateKey` bindings depend on the https://plugins.jenkins.io/credentials/[Credentials Plugin]
```

to retrieve user's credential using the Credentials API and https://plugins.jenkins.io/credentials-binding/[Credentials Binding Plugin]
to bind Git specific environment variables with script(sh,bat,powershell) / terminal commands depending upon the current CLI-git version installed on

> system which is being used to perform git operation that requires authentication.

Reviewer: This is a very long sentence that should be split into multiple sentences.

We prefer sentence per line in files like this.  When we write with a sentence per line, it is easier to review and comment and is easier to see when a sentence is too long.

> Both the bindings works on behalf of the user to provide git authentication support, without their interaction with the command-line.

Reviewer: ```suggestion
Both the bindings works on behalf of the user to provide git authentication support, without complicated steps to pass the credentials to command line git.
```

> The bindings only provide authentication support for the `Git CLI` implementation.

Reviewer: ```suggestion
The bindings only provide authentication support for the command line git implementation.
```

> Other Git implementations inside the `git client` plugin such as `JGit` and `JGit with Apache HTTP Client` are not supported.

Reviewer: ```suggestion
Other Git implementations inside the git client plugin such as `JGit` and `JGit with Apache HTTP Client` are not supported.
```

> To facilitate this limitation users are provided with a dropbox which lists all the Git-Tool implementation of

Reviewer: ```suggestion
Users are provided with a dropbox which lists all the Git-Tool implementation of
```

CLI-Git configured in Jenkins `Global System Configuration`.

NOTE:: If no CLi-Git implemented Git-Tool is available in `Global System Configuration` then by default JGit
implemented Git-Tool will be used.

==== Git Username and Password Binding

The `gitUsernamePassword` implementation uses the Jenkins username and password
values retrieved through Credential API, to access a remote repository  over *HTTP* protocol.
The binding uses the `GIT_ASKPASS` environment variable to provide credentials requested by a git operation in a pipeline job/freestyle project.

> Executable script(sh,bat,powershell) is attached/binded to `GIT_ASKPASS` variable, which is invoked when asked for user’s credentials by HTTP server.

Reviewer: ```suggestion
Executable script(sh,bat,powershell) is bound to the `GIT_ASKPASS` variable, which is invoked when asked for the user’s credentials by HTTP server.
```

Pipeline Job:: Using Git Username and Password Binding in a Pipeline Job

> image::../../../../images/gsoc/2021/git-credentials/usernamepassword-binding-pipline-job.png[Git Username and Password in Pipeline Job]

Reviewer: Don't need the relative path here
```suggestion
image:/images/gsoc/2021/git-credentials/usernamepassword-binding-pipline-job.png[Git Username and Password in Pipeline Job]
```

Pipeline Job:: Using Git Username and Password Binding in a FreeStyle-Project

> image::../../../../images/gsoc/2021/git-credentials/usernamepassword-binding-freestyle-project.png[Git Username and Password in FreeStyle Project]

Reviewer: Don't need relative path here
```suggestion
image:/images/gsoc/2021/git-credentials/usernamepassword-binding-freestyle-project.png[Git Username and Password in Freestyle Project]
```

> IMPORTANT: By default, two variable bindings `GIT_USERNAME` and `GIT_PASSWORD`,
> bind with user's credential i.e. username and password respectively.

Reviewer: ```suggestion
IMPORTANT: Two variable bindings `GIT_USERNAME` and `GIT_PASSWORD` provide the username and password respectively.
```

These variable bindings are also accessible when using `JGit/JGit with Apache HTTP Client` Git-Tool

> implementations in both freestyle project and pipeline job.

Reviewer: ```suggestion
implementations in both freestyle jobs and pipeline jobs.
```

Author: Is it Freestyle project or Freestyle job, I have always used Freestyle project.

Reviewer: The words "job" and "project" are synonyms in Jenkins.  My issue was having a single sentence that uses both words.  I feared that someone might not realize that "job" and "project" are synonyms in Jenkins.  Just checked the [glossary](https://www.jenkins.io/doc/book/glossary/) and it says that "job" is a deprecated synonym for "project".  So, my change is in the wrong direction.  It should have changed "job" to "project" rather than changing "project" to "job"

==== Git SSH Private Key Binding

The `gitSshPrivateKey` implementation provides git authentication support over *SSH* protocol
using private key and passphrase credentials of a user. The binding uses two git specific environment
variables depending upon the minimum CLI-git version

> ** `GIT_SSH_COMMAND` - If version is greater than 2.3, then the GIT_SSH_COMMAND environment variable provides ssh command including the

Reviewer 2: provides the ssh command

> ** `GIT_SSH_COMMAND` - If version is greater than 2.3, then the GIT_SSH_COMMAND environment variable provides ssh command including the

Reviewer: ```suggestion
** `GIT_SSH_COMMAND` - If version is greater than 2.3, then the GIT_SSH_COMMAND environment variable provides the ssh command including the
```

> necessary options which are: path to the private key and host key checking to authenticate and connect git server without using an executable script.

Reviewer 2: and connect to the git server

> ** `SSH_ASKPASS` - If version is less than 2.3, an executable script is attached to the variable which provides

Reviewer 2: One of these is valid when version 2.3 is used, which one? You have "greater than" and "less than" but no "equal to".

> ssh command including the necessary options which are: path to the private key and host key checking to authenticate and connect git server

Reviewer 2: Missing some articles:

the ssh command
connect to the git server

Support for private key formats

** The following key formats are supported through Jenkins plugins

* `BouncyCastle API Plugin` - Supports decryption of PKCS#8 PEM encoded and PEM private keys.
* `SSHD Plugin` - Provides a transitive dependency i.e. Apache SSHD-Core v2.7.0, which supports decryption of
OpenSSH/RFC4716 formatted private keys.

Support for encryption algorithms

* The following encryption algorithms are supported in various formats discussed above

** RSA
** DSA
** ECDSA
** ED25519

===== Decryption of private keys

> The passphrase protected private key needs to be decrypted before they could be used to perform

Reviewer 2: before it could be used (I think the subject is "The passphrase" which is singular not plural).

> The passphrase protected private key needs to be decrypted before they could be used to perform

Reviewer: ```suggestion
The passphrase protected private key needs to be decrypted before it can be used to perform
```

Reviewer: This is a very long sentence that should be split into multiple sentences.

> git authentication operation over *SSH* protocol, if not decrypted then a prompt is displayed to

Reviewer 2: Articles missing:
the git authentication operation over the *SSH* protocol.

> user asking for the passphrase of the private key being used in the current job/project, which is not a

Reviewer 2: Article missing:
the user

> required behaviour for this binding.

Reviewer 2: I do not understand what you mean by "which is not a required behavior for this binding". Which binding are we talking about? I though this was about the private key decryption. I guess I am not familiar with this topic.

> To tackle this issue, the BouncyCastle API plugin and SSHD plugin are used which provide decryption support

Reviewer 2: are used which provide -> are used to provide

> for various private formats mentioned above.

Reviewer 2: for various -> for the various

Pipeline Job:: Using Git SSH Private Key in a Pipeline Job

> image::../../../../images/gsoc/2021/git-credentials/ssh-private-key-pipeline-job.png[Git SSH Private Key in Pipeline Job]

Reviewer: ```suggestion
image:/images/gsoc/2021/git-credentials/ssh-private-key-pipeline-job.png[Git SSH Private Key in Pipeline Job]
```

Pipeline Job:: Using Git SSH Private Key Binding in a FreeStyle-Project

> image::../../../../images/gsoc/2021/git-credentials/ssh-private-key-freestyle-job.png[Git SSH Private Key in FreeStyle Project]

Reviewer: ```suggestion
image:/images/gsoc/2021/git-credentials/ssh-private-key-freestyle-job.png[Git SSH Private Key in FreeStyle Project]
```
